bip-0000.mediawiki

```diff
@@ -0,0 +1,335 @@
+* Inputs used to derive the shared secret are from the ''[[#inputs-for-shared-secret-derivation|Inputs For Shared Secret Derivation]]'' list
+
+A transaction is not a ''Silent Payments v0'' transaction and should be skipped entirely<ref name="why_skip_transactions">'''Why skip transactions that spend unknown output scripts?''' Skipping transactions that spend unknown output scripts allows us to have a clean upgrade path for Silent Payments by avoiding the need to scan the same transaction multiple times with different rule sets. If a fancy new output type is added in the future and Silent Payments v1 is released with support, we would want to avoid having to first scan the transaction with the silent payment v0 rules and then again with the silent payment v1 rules.</ref> when scanning if any of the following are true:
+
+* The transaction does not contain any BIP341 taproot outputs
+* The transaction does not have at least one input from the ''[[#inputs-for-shared-secret-derivation|Inputs For Shared Secret Derivation]]'' list
+* The transaction inputs contain a new, undefined output type (e.g SegWit versions > 1)
+
+=== Address encoding ===
+
+A silent payment address is constructed in the following manner:
+
+* Let ''B<sub>scan</sub>, b<sub>scan</sub> = Receiver’s [https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki BIP340] scan public key and corresponding private key''
+* Let ''B<sub>spend</sub>, b<sub>spend</sub> = Receiver’s [https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki BIP340] spend public key and corresponding private key''
+* Let ''B<sub>m</sub> = B<sub>spend</sub> + m·G'', where ''m'' an optional integer tweak for labeling
+** In the case of ''m'' = 0, no label is applied and ''B<sub>m</sub> = B<sub>spend</sub>''
+* The final address is a [https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki Bech32m] encoding of:
+** The human-readable part "sp" for mainnet, "tsp" for testnets (e.g  signet, testnet, regtest)
+** The data-part values:
+*** The character "q", to represent a silent payment address of version 0
+*** The 64 byte concatenation of the receiver's public keys, ''B<sub>m</sub> || B<sub>scan</sub>''
+
+=== Outpoints hash ===
+
+The sender and receiver MUST calculate an outpoints hash for the transaction in the following manner:
+
+* Collect each ''outpoint'' used as an input to the transaction
+* Let ''outpoints = outpoint<sub>0</sub> || … || outpoint<sub>n</sub>'', sorted by txid and vout, ascending order
```

Comment:
This makes it so the order of the inputs doesn't matter for the resulting address, which is slightly safer even if issues around this are unlikely to occur. For instance, a wallet might RBF bump a tx and for some reason decide the change the order of the inputs, and neglect to realize this also means they'd have to recalculate the SP output. Now they won't have to.

Reply:
> order defined by the transaction

In addition to Ruben's comment, I'd add I don't think this order is defined until the transaction is signed, whereas you are generating the silent payment output before signing. Seems much safer to have an ordering defined by the protocol rather than requiring wallet software to ensure the prevout outpoints are always kept in the same order during the entire transaction creation process

Reply:
Thank you both for comments. In case you'd like to turn them into a footnote, it could be something like:

**Why are outpoints sorted before hashing?** This way the silent payment output does not need to be recalculated if the wallet changes the order of inputs, e.g. at signing time or during a RBF bump.

Reply:
Thanks for the suggested footnote! I've added it.